### DISCUSSION TITLE
Splitting gc.collect() in small chunks

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#11015] [General]
Hi,
Is there a way to run gc.collect() in small chunks ?
I have an application that shall run at 100hz.
gc.collect is run manually and takes about 20ms. Even when run at every cycles, gc.collect duration does not decreases below these 20ms.
I would like to reduce this 20ms duration by splitting gc.collect in smaller pieces, then running these small pieces on every application cycles.
Thanks for pointers

**Comment by rkompass** [comment]
Have a look at [https://github.com/micropython/micropython/pull/10235](https://github.com/micropython/micropython/pull/10235).

**Reply by AutoPilotjn** [reply]
Thanks.
Looks very interesting. Has the PR been integrated in the nightly builds ?

**Reply by rkompass** [reply]
Apparently not. Could you try it out and compare with your measurements?

**Reply by AutoPilotjn** [reply]
On an 1.19.1 local cloned release, I have copied gc.c and mpstate.h from https://github.com/micropython/micropython/pull/10235 in py directory, then recompiled (doing nothing more). With or without a gc.threshold(150_000), no differences with both damz & 1.19.1 releases:

damz improvement

```
Before gc. stats: mem free: 105872 - mem_alloc: 165776
After gc. stats: mem free: 171312 - mem_alloc: 100336
Time consumed by gc(us): 025781, cycle: 0-17
```

MicroPython 1.19.1
```
Before gc. stats: mem free: 105440 - mem_alloc: 166208
After gc. stats: mem free: 171312 - mem_alloc: 100336
Time consumed by gc (us): 025781, cycle: 0-17
```

My app has  about 100.000 bytes of permanently allocated memory. How can the gc duration runs be decreased ? 
I open a new thread https://github.com/micropython/micropython/discussions/11029 for general questions regarding memory allocation management.

**Reply by rkompass** [reply]
This is in contradiction to the statements and measurements in [https://github.com/micropython/micropython/pull/10235](https://github.com/micropython/micropython/pull/10235).
About 70 K might be too much memory to be freed though. Did you try the PR in a task that had less memory garbaged?
Otherwise there might be an error somewhere.

**Reply by rkompass** [reply]
Also a thorough search on the formerly used [Forum](https://forum.micropython.org) and on GitHub should yield more contributions. I remember other improvements of GC being presented and discussed.
